commit/merkleroot/rmn/client.go

```diff
@@ -19,13 +33,533 @@ type Client interface {
+type PBClient struct {
+	lggr                                    logger.Logger
+	rawRmnClient                            RawRmnClient
+	rmnNodes                                []RMNNodeInfo
+	rmnRemoteAddress                        []byte
+	rmnHomeConfigDigest                     []byte
+	minObservers                            int
+	minSigners                              int
+	observationsInitialRequestTimerDuration time.Duration
+	reportsInitialRequestTimerDuration      time.Duration
+}
+
+// RMNNodeInfo contains the information about an RMN node.
+type RMNNodeInfo struct {
+	ID                    uint32 // ID is the index of this node in the RMN config
+	SupportedSourceChains mapset.Set[uint64]
+	IsSigner              bool
+}
+
+// NewPBClient creates a new RMN Client to be used by the plugin.
+func NewPBClient(
+	lggr logger.Logger,
+	rawRmnClient RawRmnClient,
+	rmnNodes []RMNNodeInfo,
+	rmnRemoteAddress []byte,
+	rmnHomeConfigDigest []byte,
+	minObservers int,
+	minSigners int,
+	observationsInitialRequestTimerDuration time.Duration,
+	reportsInitialRequestTimerDuration time.Duration,
+) *PBClient {
+	return &PBClient{
+		lggr:                                    lggr,
+		rawRmnClient:                            rawRmnClient,
+		rmnNodes:                                rmnNodes,
+		rmnRemoteAddress:                        rmnRemoteAddress,
+		rmnHomeConfigDigest:                     rmnHomeConfigDigest,
+		minObservers:                            minObservers,
+		minSigners:                              minSigners,
+		observationsInitialRequestTimerDuration: observationsInitialRequestTimerDuration,
+		reportsInitialRequestTimerDuration:      reportsInitialRequestTimerDuration,
+	}
+}
+
+func (c *PBClient) ComputeReportSignatures(
+	ctx context.Context,
+	destChain *rmnpb.LaneDest,
+	requestedUpdates []*rmnpb.FixedDestLaneUpdateRequest,
+) (*ReportSignatures, error) {
+	// Group the lane update requests by their source chain.
+	lursPerChain := make(map[uint64]lurWithMeta)
+	for _, updateReq := range requestedUpdates {
+		if _, exists := lursPerChain[updateReq.LaneSource.SourceChainSelector]; exists {
+			return nil, errors.New("this Client implementation assumes each lane update is for a different chain")
+		}
+
+		lursPerChain[updateReq.LaneSource.SourceChainSelector] = lurWithMeta{
+			lur:      updateReq,
+			rmnNodes: mapset.NewSet[uint32](),
+		}
+		for _, node := range c.rmnNodes {
+			if node.SupportedSourceChains.Contains(updateReq.LaneSource.SourceChainSelector) {
+				lursPerChain[updateReq.LaneSource.SourceChainSelector].rmnNodes.Add(node.ID)
+			}
+		}
+	}
+
+	// Filter out the lane update requests for chains without enough RMN nodes supporting them.
+	for chain, l := range lursPerChain {
+		if l.rmnNodes.Cardinality() < c.minObservers {
+			delete(lursPerChain, chain)
+		}
+	}
+
+	if len(lursPerChain) == 0 {
+		return nil, errors.New("no source chains with enough RMN nodes, nothing to do")
```

Comment:
Same here, only on misconfigs right?

Might be useful to log the exact misconfiguration. For example, I think this would only happen if:

* RMN is enabled, AND
* `minObservers` is not set correctly on ALL chains

Is this correct? If so might be useful to include in the error (which will show up in a log) making it easy to diagnose the situation.

Reply:
done